### PR TITLE
Tab switch crash bug

### PR DESCRIPTION
**Pull request by SunnyNagam**
Tabs now load only when opened and stay loaded and won't make redundant api calls. via manual tab caching

### CODE REVIEW

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt [hunk 1]
```diff
@@ -24,27 +26,44 @@ import timber.log.Timber
 
 class TabFragment(index: Int) : Fragment() {
     var tabIndex = index
+    var articles : List<Article>? = null
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt:29]: Should be `val` I believe

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt [hunk 2]
```diff
@@ -24,27 +26,44 @@ import timber.log.Timber
...
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+        if(activity != null){               // Afterwards keep displaying already initialized articles
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt:33]: I think our convention should be to not use in-line comments, would rather it was on top of the code block its trying to explain

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt [hunk 3]
```diff
@@ -24,27 +26,44 @@ import timber.log.Timber
...
-
         val retrofit = Retrofit.Builder()
             .baseUrl("https://factfulnews.herokuapp.com")
             .addConverterFactory(GsonConverterFactory.create())
             .build()
 
-        val api = retrofit.create(ApiService::class.java)
-        api.fetchArticles(CategoryUtils.categories[tabIndex]).enqueue(object: Callback<FetchResponse> {
-            override fun onResponse(call: Call<FetchResponse>, response: Response<FetchResponse>) {
-                showArticles(response.body()!!.articles)
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt]: I notice this PR needs to rebase on top of your own previous commit LOL https://github.com/Cobresun/factfulnewsAndroid/pull/13 also one that I put in, so make sure to rebase and then double check everything works ...

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt [hunk 4]
```diff
@@ -24,27 +26,44 @@ import timber.log.Timber
...
+                        Toast.makeText(context, "Error Fetching response!", Toast.LENGTH_SHORT).show()
+                    }
+                }
 
-            override fun onFailure(call: Call<FetchResponse>, t: Throwable) {
-                Timber.d(t.toString()) // Log the failure.
-            }
-        })
+                override fun onFailure(call: Call<FetchResponse>, t: Throwable) {
+                    Timber.d(t.toString()) // Log the failure.
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt]: I just noticed how redundant this comment is, I think we can get rid of it 👍

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabsPagerAdapter.kt [hunk 5]
```diff
@@ -8,8 +8,14 @@ class TabsPagerAdapter(fragmentManager: FragmentManager, numCategories: Int):
...
 
+    val tabsCache : Array<TabFragment?> = arrayOfNulls<TabFragment?>(numCategories)
+
     override fun getItem(position: Int): Fragment {
-        return TabFragment(position)
+        if(tabsCache[position] == null){
+            tabsCache[position] = TabFragment(position)
+        }
+
+        return tabsCache[position] ?: TabFragment(position)
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/TabsPagerAdapter.kt:15]: Sick use of the Elvis operator! 😎

File: app/build.gradle [hunk 6]
```diff
@@ -10,8 +10,8 @@ android {
         applicationId "com.cobresun.factfulnewsandroid"
         minSdkVersion 23
         targetSdkVersion 28
-        versionCode 1
-        versionName "1.0"
+        versionCode 3
+        versionName "1.2"
```
Comment (brian-norman) [inline app/build.gradle:14]: Much better to have the version upgrades be separate PRs so we can see them in the commit log

File: app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt [hunk 7]
```diff
@@ -31,10 +31,10 @@ class MainActivity : AppCompatActivity() {
         // Each page is represented by its own fragment.
         val viewPager = findViewById<ViewPager>(R.id.pager)
 
-        var adapter: TabsPagerAdapter
-        adapter = TabsPagerAdapter(supportFragmentManager, tabLayout.tabCount)
+        viewPager.adapter = TabsPagerAdapter(supportFragmentManager, tabLayout.tabCount)
+
+        viewPager.offscreenPageLimit = 0
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt]: Whats the argument for changing this? Leaving it at default allows for smoother tab switching doesn't it?

**Review by brian-norman — CHANGES_REQUESTED** [review]
Nice PR, had some comments, but mostly looks clean!

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt (hunk 1 above)
Comment (SunnyNagam) [reply app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt:29]: nope it needs to reassigned i think

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt (hunk 3 above)
Comment (SunnyNagam) [reply app/src/main/java/com/cobresun/factfulnewsandroid/TabFragment.kt]: oh yikes

File: app/src/main/java/com/cobresun/factfulnewsandroid/TabsPagerAdapter.kt (hunk 5 above)
Comment (SunnyNagam) [reply app/src/main/java/com/cobresun/factfulnewsandroid/TabsPagerAdapter.kt:15]: 🎉

File: app/build.gradle (hunk 6 above)
Comment (SunnyNagam) [reply app/build.gradle:14]: noted for future but for now this updated version has aldready been uploaded to play

File: app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt (hunk 7 above)
Comment (SunnyNagam) [reply app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt]: true i just left that in by accident when i was testing something. Fun fact tho, apparently setting it to 0 changes nothing.

Comment (brian-norman) [reply app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt]: Why do we still have this then ?!

File: app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt [hunk 8]
```diff
@@ -33,6 +33,9 @@ class MainActivity : AppCompatActivity() {
 
         viewPager.adapter = TabsPagerAdapter(supportFragmentManager, tabLayout.tabCount)
 
+        viewPager.offscreenPageLimit = 0
+
+
```
Comment (brian-norman) [inline app/src/main/java/com/cobresun/factfulnewsandroid/ui/activities/MainActivity.kt]: Unnecessary blank spaces I think

**Review by brian-norman — CHANGES_REQUESTED** [review]

**Review by brian-norman — APPROVED** [review]
